racket/collects/racket/match/struct.rkt

```diff
@@ -17,57 +34,43 @@
-         (for-each (lambda (an)
-                     (syntax-case an ()
-                       [(field pat)
-                        (unless (identifier? #'field)
-                          (raise-syntax-error 
-                           'struct* "not an identifier for field name" 
-                           stx #'field))
-                        (let ([field-acc
-                               (datum->syntax #'field
-                                              (string->symbol
-                                               (format "~a-~a"
-                                                       (syntax-e #'struct-name)
-                                                       (syntax-e #'field)))
-                                              #'field)])
-                          (when (free-id-table-ref field-acc->pattern field-acc #f)
-                            (raise-syntax-error 'struct* "Field name appears twice" stx #'field)) 
-                          (free-id-table-set! field-acc->pattern field-acc #'pat))]
-                       [_
-                        (raise-syntax-error
-                         'struct* "expected a field pattern of the form (<field-id> <pat>)"
-                         stx an)]))
-                   (syntax->list #'(field+pat ...)))
-         (let* (; Get the structure info
-                [acc (fourth (extract-struct-info v))]
-                ;; the accessors come in reverse order
-                [acc (reverse acc)]
-                ;; remove the first element, if it's #f
-                [acc (cond [(empty? acc) acc]
-                           [(not (first acc)) (rest acc)]
-                           [else acc])]
-                ; Order the patterns in the order of the accessors
-                [pats-in-order
-                 (for/list ([field-acc (in-list acc)])
-                   (begin0
-                     (free-id-table-ref
-                      field-acc->pattern field-acc
-                      (syntax/loc stx _))
-                     ; Use up pattern
-                     (free-id-table-remove! field-acc->pattern field-acc)))])
-           ; Check that all patterns were used
-           (free-id-table-for-each
-            field-acc->pattern
-            (lambda (field-acc pat)
-              (when pat
-                (raise-syntax-error 'struct* "field name not associated with given structure type"
-                                    stx field-acc))))
-           (quasisyntax/loc stx
-             (struct struct-name #,pats-in-order))))])))
+         (define the-struct-info (extract-struct-info v))
+
+         ;; own-fields and all-accessors are in the reverse order
+         (define all-accessors (list-ref the-struct-info 3))
+         (define own-fields
+           (if (struct-field-info? v)
+               (struct-field-info-list v)
+               (extract-field-names stx the-struct-info)))
+         (define field-set (list->set own-fields))
```

Comment:
So that checking for membership could be fast. Is there an alternative way to do it?

Reply:
I expect that the conversion cost dominates the checking cost for any plausible number of fields.

Reply:
Well, I know one Racket project that generates thousands of struct fields (see https://github.com/emina/rosette/issues/147). This used to trigger a really long compilation time, and IIRC, @mflatt fixed the issue was it asked in Slack (see https://github.com/racket/racket/commit/0de88f203d53c159dc968f65ee1c3d33e94ece3c for the fix). 

Using a set is meant to avoid quadratic time, but I agree with you that for most struct definitions, it is overkill. 

One possible solution (which unfortunately would complicate the code more) is to use either set or list membership depending on the number of fields, with the threshold at, say, 50 fields. Does this sound reasonable?

Reply:
Ok, that makes sense. I think simplicity is the right approach here.

Reply:
Does using `racket/set` affect the footprint of match noticeably, given that `racket/set` includes a lot of extra geenrality and syntax and contracts? An alternative is to use a hash table mapping keys to `#t`.

Reply:
I did experiment `(require racket/match)` before and after this change but didn't find significant difference. That being said, I think a hash map mapping to `#t` is better. I will submit a change soon.